rpc/json-rpc/src/response.rs

```diff
@@ -0,0 +1,402 @@
+impl<T> std::fmt::Display for Response<T>
+where
+    T: Clone + Serialize,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match serde_json::to_string_pretty(self) {
+            Ok(json) => write!(f, "{json}"),
+            Err(_) => Err(std::fmt::Error),
+        }
+    }
+}
+
+//---------------------------------------------------------------------------------------------------- Key
+/// This type represents the `"key"` values within [`Response`].
+///
+/// We need this since `Response` has a custom deserializer implementation.
+pub(crate) enum Key {
+    /// "jsonrpc" field.
+    JsonRpc,
+    /// "result" field.
+    Result,
+    /// "error" field.
+    Error,
+    /// "id" field.
+    Id,
+}
+
+/// Serde [`Key`] visitor marker struct.
+struct KeyVisitor;
+
+impl serde::de::Visitor<'_> for KeyVisitor {
+    type Value = Key;
+
+    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+        formatter.write_str("Key field must be a string and one of the following values: ['jsonrpc', 'result', 'error', 'id']")
+    }
+
+    fn visit_str<E: serde::de::Error>(self, text: &str) -> Result<Self::Value, E> {
+        // FIXME: I don't think JSON-RPC 2.0 specifies
+        // case-sensitiveness so just be strict here.
+        Ok(match text {
+            "jsonrpc" => Key::JsonRpc,
+            "result" => Key::Result,
+            "error" => Key::Error,
+            "id" => Key::Id,
+            _ => {
+                return Err(serde::de::Error::invalid_value(
+                    serde::de::Unexpected::Str(text),
+                    &self,
+                ))
+            }
+        })
+    }
+}
+
+impl<'a> Deserialize<'a> for Key {
+    fn deserialize<D: Deserializer<'a>>(des: D) -> Result<Self, D::Error> {
+        des.deserialize_str(KeyVisitor)
+    }
+}
+
+//---------------------------------------------------------------------------------------------------- Serde impl
+impl<T> Serialize for Response<T>
+where
+    T: Serialize + Clone,
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut s = serializer.serialize_struct("Response", 3)?;
+
+        s.serialize_field("jsonrpc", &self.jsonrpc)?;
+
+        // This member is required.
+        //
+        // Even if `null`, or the client `Request` didn't include one.
+        s.serialize_field("id", &self.id)?;
+
+        match &self.payload {
+            Ok(r) => s.serialize_field("result", r)?,
+            Err(e) => s.serialize_field("error", e)?,
+        }
+
+        s.end()
+    }
+}
+
+// [`Response`] has a manual deserialization implementation because
+// we need to confirm `result` and `error` don't both exist:
+//
+// > Either the result member or error member MUST be included, but both members MUST NOT be included.
+//
+// <https://www.jsonrpc.org/specification#error_object>
+impl<'de, T> Deserialize<'de> for Response<T>
```

Comment:
Not sure if you can do this with `serde` macros. Manually implementing is painful.

Reply:
Is this here just because we can't have both `result` and `error`? if so, wouldn't it be possible to solve this with our own `Result` type that renames it's `Ok` to `result` and `Err` to `error`?

Reply:
Replaced with `enum ResponsePayload`.